[DR-3210] Sign GCP urls for parquet exports

nmalfroy opened the pull request:
We've run into a bunch of issues with permission propagation in Google when importing TDR exports.  This PR changes the default behavior of exports to provide signed URLs for the manifest and all parquet files instead of `gs://` paths.  These signed URLs are immediately accessible by the import-service and as of the next deployment, the import-service will be capable of ingesting these.

A couple of notes:
- The original behavior was left in but the default behavior is now to use signed URLs
- The length of the signed URL is 60 minutes, which seems like a reasonable amount of time to ingest

nmalfroy committed:
Add flight support for optionally signing URLs for GCP snapshot exports

nmalfroy committed:
Add service and controller level code to allow exports using signed URLs

nmalfroy committed:
Update integration test to validate exports using signed urls

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/flight/export/SnapshotExportFlight.java:
```diff
@@ -51,6 +53,10 @@ public SnapshotExportFlight(FlightMap inputParameters, Object applicationContext
             inputParameters.get(ExportMapKeys.EXPORT_VALIDATE_PK_UNIQUENESS, Boolean.class), true);
 
+    boolean signUrls =
+        Objects.requireNonNullElse(
+            inputParameters.get(ExportMapKeys.EXPORT_SIGN_URLS, Boolean.class), true);
```
Under what conditions would this be null? Is this to handle the case where a flight is launched by TDR version X and then is run by TDR version Y? Otherwise it seems like it should always be set, and if it it isn't, it's a server error.

nmalfroy replied:
Yeah, this is really for that intermediary race condition

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/flight/export/SnapshotExportWriteManifestStep.java:
```diff
@@ -65,25 +80,32 @@ public StepResult doStep(FlightContext context) throws InterruptedException, Ret
+                                ? entry.getValue().stream()
+                                    .map(path -> signGsUrl(storage, path))
+                                    .toList()
+                                : entry.getValue()))
```
Not sure it's much better but you could move the `?` inside the inner `map`, to save a line of code.
```suggestion
                                entry.getValue().stream()
                                    .map(path -> signUrls ? signGsUrl(storage, path) : path)
                                    .toList()
```

nmalfroy replied:
Yeah, I'm not sure which I like better.  Kinda 50/50 so I'll just leave

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/flight/export/SnapshotExportWriteManifestStep.java:
```diff
@@ -65,25 +80,32 @@ public StepResult doStep(FlightContext context) throws InterruptedException, Ret
+            .toList();
 
+    String effectiveExportManifestFile =
+        signUrls ? signGsUrl(storage, exportManifestPath) : exportManifestPath;
```
Since it's done in two places (and `storage` is the same everywhere) you could create a local lambda for this operation.

```
    Storage storage = gcsProjectFactory.getStorage(snapshot.getDataProject(), true);
    UnaryOperator<String> maybeSignUrl = signUrls ? (path) -> signGsUrl(storage, path) : path -> path;
[...]
                entry ->
                    new SnapshotExportResponseModelFormatParquetLocationTables()
                        .name(entry.getKey())
                        .paths(entry.getValue().stream().map(maybeSignUrl).toList()))
[...]
                        new SnapshotExportResponseModelFormatParquet()
                            .manifest(maybeSignUrl.apply(exportManifestPath))
                            .location(
                                new SnapshotExportResponseModelFormatParquetLocation()
                                    .tables(tables))));
```

nmalfroy replied:
Hmm, yeah.  Personal preference, I'm finding that one a little harder to read

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/flight/export/SnapshotExportWriteManifestStep.java:
```diff
@@ -107,4 +129,18 @@ public StepResult doStep(FlightContext context) throws InterruptedException, Ret
     return StepResult.getStepResultSuccess();
   }
+
+  private String signGsUrl(Storage storage, String gsPath) {
```
Maybe this should be in `GcsUriUtils`? It seems general and not directly related to `SnapshotExportWriteManifestStep`.

nmalfroy replied:
Yeah, I've been torn because there are a couple of different ways that we sign urls.  But yeah, that's fair.

nmalfroy replied:
I created this simple signing method but I held off on migrating the other uses since those a bit more involved for now (have to deal with requester pays, etc.)

pshapiro4broad commented on src/test/java/bio/terra/service/snapshot/SnapshotExportIntegrationTest.java:
```diff
@@ -234,19 +259,46 @@ public void snapshotExportTest() throws Exception {
+          notAuthorizedException = ex;
+        }
+        // Appeasing SpotBugs
+        assert notAuthorizedException != null;
```
Could `assertThrows` be used here? That should also make the linter happy, and avoid writing code that's partly copying what `assertThrows` does.

Also note that an assertion statement has no effect unless you run using `-enableassertions`. If you want it to do something, a function call is required.

nmalfroy replied:
yeah good call

pshapiro4broad commented on src/test/java/bio/terra/service/snapshot/SnapshotExportIntegrationTest.java:
```diff
@@ -234,19 +259,46 @@ public void snapshotExportTest() throws Exception {
-      } catch (StorageException ex) {
-        notAuthorizedException = ex;
+  private byte[] readUrl(String url) throws IOException {
+    try (CloseableHttpClient client = HttpClients.createDefault()) {
+      HttpUriRequest request = new HttpGet(url);
+      try (CloseableHttpResponse response = client.execute(request); ) {
+        assertThat(
+            "Signed URL is accessible",
+            response.getStatusLine().getStatusCode(),
+            equalTo(HttpStatus.OK.value()));
+
+        byte[] bytes;
+        try (InputStream responseBody = response.getEntity().getContent()) {
```
Is there a reason to do it this way? You can read directly from a `URL` using `URL.openStream()`.
```
  private byte[] readUrl(String url) throws IOException {
    try (var responseBody = new URL(url).openStream()) {
      return responseBody.readAllBytes();
    }
  }
```

nmalfroy replied:
Oh interesting. I was doing this the way that we've done it in other places where we do a `HEAD` on the file.  Lemme use this instead

nmalfroy replied:
Actually, I'm going to leave as is since my way explicitly checks for a successful status (whereas presumably I could end up having a 400 or 403 back here and this would just return the error body)

pshapiro4broad replied:
You can try and see, but I'm pretty sure a 400 or 403 would result in a thrown `IOException`. Looking at the source, any status >=400 results in an `IOException`. https://github.com/openjdk/jdk17/blob/master/src/java.base/share/classes/sun/net/www/protocol/http/HttpURLConnection.java#L1991C3-L1991C3

nmalfroy replied:
Ah fair enough.  I'll use this then!

pshapiro4broad reviewed (approved):
Looks good, a few comments. Also I was wondering about the branch name, as it's a PR for gcp export not parquet. Maybe it's exporting _to_ parquet?

nmalfroy commented:
> Looks good, a few comments. Also I was wondering about the branch name, as it's a PR for gcp export not parquet. Maybe it's exporting _to_ parquet?

Yeah, this is parquet on GCP

nmalfroy committed:
PR Feedback

nmalfroy committed:
Don't sign urls for tests (where it doesn't matter)

It doesn't matter since this is just testing the contents of the parquet file, not the actual pointers to the parquet
